File: apollo-runtime/src/main/java/com/apollographql/android/impl/ApolloParseInterceptor.java
package com.apollographql.android.impl;

import com.apollographql.android.CustomTypeAdapter;
import com.apollographql.android.api.graphql.Operation;
import com.apollographql.android.api.graphql.Response;
import com.apollographql.android.api.graphql.ResponseFieldMapper;
import com.apollographql.android.api.graphql.ScalarType;
import com.apollographql.android.cache.http.HttpCache;
import com.apollographql.android.cache.normalized.ResponseNormalizer;
import com.apollographql.android.impl.util.HttpException;
import com.apollographql.android.internal.ApolloLogger;

import java.io.Closeable;
import java.io.IOException;
import java.util.Map;
import java.util.concurrent.ExecutorService;

import javax.annotation.Nonnull;

final class ApolloParseInterceptor implements ApolloInterceptor {
  private final HttpCache httpCache;
  private final ResponseNormalizer<Map<String, Object>> normalizer;
  private final ResponseFieldMapper responseFieldMapper;
  private final Map<ScalarType, CustomTypeAdapter> customTypeAdapters;
  private final ApolloLogger logger;

  ApolloParseInterceptor(HttpCache httpCache, ResponseNormalizer<Map<String, Object>> normalizer,
      ResponseFieldMapper responseFieldMapper, Map<ScalarType, CustomTypeAdapter> customTypeAdapters,
      ApolloLogger logger) {
    this.httpCache = httpCache;
    this.normalizer = normalizer;
    this.responseFieldMapper = responseFieldMapper;
    this.customTypeAdapters = customTypeAdapters;
    this.logger = logger;
  }

  @Override @Nonnull public InterceptorResponse intercept(Operation operation, ApolloInterceptorChain chain)
      throws IOException {
    InterceptorResponse response = chain.proceed();
    return parse(operation, response.httpResponse.get());
  }

  @Override
  public void interceptAsync(@Nonnull final Operation operation, @Nonnull ApolloInterceptorChain chain,
      @Nonnull ExecutorService dispatcher, @Nonnull final CallBack callBack) {
    chain.proceedAsync(dispatcher, new CallBack() {
      @Override public void onResponse(@Nonnull InterceptorResponse response) {
        try {
          callBack.onResponse(parse(operation, response.httpResponse.get()));
        } catch (Exception e) {
          callBack.onFailure(e);
        }
      }

      @Override public void onFailure(@Nonnull Throwable t) {
        callBack.onFailure(t);
      }
    });
  }

  @Override public void dispose() {
    //no op
  }

  @SuppressWarnings("unchecked") private InterceptorResponse parse(Operation operation, okhttp3.Response
      httpResponse) throws IOException {
    String cacheKey = httpResponse.request().header(HttpCache.CACHE_KEY_HEADER);
    if (httpResponse.isSuccessful()) {
      try {
        HttpResponseBodyParser parser = new HttpResponseBodyParser(operation, responseFieldMapper, customTypeAdapters);
        Response parsedResponse = parser.parse(httpResponse.body(), normalizer);
        return new InterceptorResponse(httpResponse, parsedResponse, normalizer.records());
      } catch (Exception rethrown) {
        logger.e(rethrown, "Failed to parse network response for operation %s", operation);
        closeQuietly(httpResponse);
        if (httpCache != null) {
          httpCache.removeQuietly(cacheKey);
        }
        throw rethrown;
      }
    } else {
      throw new HttpException(httpResponse);
    }
  }

  private static void closeQuietly(Closeable closeable) {
    if (closeable != null) {
      try {
        closeable.close();
      } catch (RuntimeException rethrown) {
        throw rethrown;
      } catch (Exception ignored) {
      }
    }
  }
}

File: apollo-runtime/src/main/java/com/apollographql/apollo3/internal/interceptor/ApolloParseInterceptor.kt
package com.apollographql.apollo3.internal.interceptor

import com.apollographql.apollo3.CacheInfo
import com.apollographql.apollo3.api.Operation
import com.apollographql.apollo3.api.CustomScalarAdapters
import com.apollographql.apollo3.api.cache.http.HttpCache
import com.apollographql.apollo3.api.internal.ApolloLogger
import com.apollographql.apollo3.api.parseResponseBody
import com.apollographql.apollo3.exception.ApolloException
import com.apollographql.apollo3.exception.ApolloHttpException
import com.apollographql.apollo3.exception.ApolloParseException
import com.apollographql.apollo3.http.OkHttpExecutionContext
import com.apollographql.apollo3.interceptor.ApolloInterceptor
import com.apollographql.apollo3.interceptor.ApolloInterceptor.CallBack
import com.apollographql.apollo3.interceptor.ApolloInterceptor.FetchSourceType
import com.apollographql.apollo3.interceptor.ApolloInterceptor.InterceptorRequest
import com.apollographql.apollo3.interceptor.ApolloInterceptor.InterceptorResponse
import com.apollographql.apollo3.interceptor.ApolloInterceptorChain
import okhttp3.Headers
import okhttp3.Response
import java.io.Closeable
import java.util.concurrent.Executor

/**
 * ApolloParseInterceptor is a concrete [ApolloInterceptor] responsible for inflating the http responses into
 * models. To get the http responses, it hands over the control to the next interceptor in the chain and proceeds to
 * then parse the returned response.
 */
class ApolloParseInterceptor(private val httpCache: HttpCache?,
                             private val customScalarAdapters: CustomScalarAdapters,
                             private val logger: ApolloLogger) : ApolloInterceptor {
  @Volatile
  var disposed = false
  override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain,
                              dispatcher: Executor, callBack: CallBack) {
    if (disposed) return
    chain.proceedAsync(request, dispatcher, object : CallBack {
      override fun onResponse(response: InterceptorResponse) {
        try {
          if (disposed) return
          val result = parse(request.operation as Operation<Operation.Data>, response.httpResponse.get())
          callBack.onResponse(result)
          callBack.onCompleted()
        } catch (e: ApolloException) {
          onFailure(e)
        }
      }

      override fun onFailure(e: ApolloException) {
        if (disposed) return
        callBack.onFailure(e)
      }

      override fun onCompleted() {
        // call onCompleted in onResponse in case of error
      }

      override fun onFetch(sourceType: FetchSourceType) {
        callBack.onFetch(sourceType)
      }
    })
  }

  override fun dispose() {
    disposed = true
  }

  @Throws(ApolloHttpException::class, ApolloParseException::class)
  fun parse(operation: Operation<Operation.Data>, httpResponse: Response): InterceptorResponse {
    val cacheKey = httpResponse.request().header(HttpCache.CACHE_KEY_HEADER)
    return if (httpResponse.isSuccessful) {
      try {
        val httpExecutionContext = OkHttpExecutionContext(httpResponse)
        var parsedResponse = operation.parseResponseBody(httpResponse.body()!!.source(), customScalarAdapters)

        parsedResponse = parsedResponse.copy(
            executionContext = parsedResponse.executionContext.plus(httpExecutionContext) + CacheInfo(httpResponse.cacheResponse() != null)
        )

        if (parsedResponse.hasErrors() && httpCache != null) {
          httpCache.removeQuietly(cacheKey!!)
        }

        InterceptorResponse(httpResponse, parsedResponse)
      } catch (rethrown: Exception) {
        logger.e(rethrown, "Failed to parse network response for operation: %s", operation.name())
        closeQuietly(httpResponse)
        httpCache?.removeQuietly(cacheKey!!)
        throw ApolloParseException("Failed to parse http response", rethrown)
      }
    } else {
      logger.e("Failed to parse network response: %s", httpResponse)
      throw ApolloHttpException(
          statusCode = httpResponse.code(),
          headers = httpResponse.headers().toMap(),
          message = httpResponse.body()?.string() ?: "",
          cause = null
      )
    }
  }

  companion object {
    private fun Headers.toMap(): Map<String, String> = this.names().map {
      it to get(it)!!
    }.toMap()

    private fun closeQuietly(closeable: Closeable?) {
      if (closeable != null) {
        try {
          closeable.close()
        } catch (ignored: Exception) {
        }
      }
    }
  }
}
